src/context/read-mode.tsx

```diff
@@ -0,0 +1,252 @@
+    InferDocumentLocalState,
+    InferDocumentOperation,
+    InferDocumentState,
+    IReadModeDriveServer,
+    ReadDocumentNotFoundError,
+    ReadDrive,
+    ReadDriveContext,
+    ReadDriveNotFoundError,
+    ReadDrivesListener,
+    ReadDrivesListenerUnsubscribe,
+    ReadDriveSlugNotFoundError,
+    RemoteDriveOptions,
+} from 'document-drive';
+import { Document, DocumentModel } from 'document-model/document';
+import { DocumentModelNotFoundError } from 'node_modules/document-drive/src/server/error';
+import {
+    createContext,
+    FC,
+    ReactNode,
+    useContext,
+    useEffect,
+    useMemo,
+    useState,
+} from 'react';
+import { drivesToHash } from 'src/hooks/useDocumentDrives';
+import { useUserPermissions } from 'src/hooks/useUserPermissions';
+import { logger } from 'src/services/logger';
+import { DefaultDocumentDriveServer } from 'src/utils/document-drive-server';
+
+export interface IReadModeContext extends IReadModeDriveServer {
+    readDrives: ReadDrive[];
+    setDocumentDrive(documentDrive: IReadModeDriveServer): void;
+}
+
+// decorator method to ensure server is defined before calling it
+function checkServer<
+    T extends IReadModeContext & { server: IReadModeDriveServer | undefined },
+    U extends any[],
+    R,
+>(
+    target: any,
+    propertyKey: string,
+    descriptor: TypedPropertyDescriptor<(this: T, ...args: U) => any>,
+): void {
+    const originalMethod = descriptor.value!;
+
+    descriptor.value = function (this: T, ...args: U): any {
+        if (!this.server) {
+            throw new Error('Read mode document drive not initialized.');
+        }
+        return originalMethod.apply(this, args) as T;
+    };
+}
+
+function bindClassMethods(instance: any) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const prototype = Object.getPrototypeOf(instance);
+    const propertyNames = Object.getOwnPropertyNames(prototype);
+
+    propertyNames.forEach(name => {
+        const descriptor = Object.getOwnPropertyDescriptor(prototype, name);
+        if (
+            descriptor &&
+            typeof descriptor.value === 'function' &&
+            name !== 'constructor'
+        ) {
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access
+            instance[name] = instance[name].bind(instance);
+        }
+    });
+}
+
+class ReadModeContextImpl implements Omit<IReadModeContext, 'readDrives'> {
+    private server?: IReadModeDriveServer;
+
+    constructor(documentDrive?: IReadModeDriveServer) {
+        bindClassMethods(this);
+        this.server = documentDrive;
+    }
+
+    getServer(): IReadModeDriveServer | undefined {
+        return this.server;
+    }
+
+    setDocumentDrive(documentDrive: IReadModeDriveServer) {
+        this.server = documentDrive;
+    }
+
+    /* eslint-disable @typescript-eslint/no-non-null-assertion */
+    @checkServer
```

Comment:
Since the DocumentDriveServer has to be set when creating the context, these methods are decorated with this method that returns an error if they are called before the server is defined.